src/Sylius/Bundle/ThemeBundle/Locator/RecursiveFileLocator.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace Sylius\Bundle\ThemeBundle\Locator;
+use Symfony\Component\Config\FileLocatorInterface;
+use Symfony\Component\Finder\Finder;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+class RecursiveFileLocator implements FileLocatorInterface
+{
+    /**
+     * @var array
+     */
+    protected $paths;
+
+    /**
+     * @param string|array $paths A path or an array of paths where to look for resources
+     */
+    public function __construct($paths = [])
+    {
+        $this->paths = (array) $paths;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function locate($name, $currentPath = null, $first = true)
+    {
+        if ('' == $name) {
+            throw new \InvalidArgumentException('An empty file name is not valid to be located.');
+        }
+
+        if ($this->isAbsolutePath($name)) {
+            if (!file_exists($name)) {
+                throw new \InvalidArgumentException(sprintf('The file "%s" does not exist.', $name));
+            }
+
+            return $name;
+        }
+
+        $directories = $this->paths;
+        if (null !== $currentPath) {
+            $directories[] = $currentPath;
+
+            $directories = array_values(array_unique($directories));
+        }
+
+        $filepaths = [];
+
+        $finder = new Finder();
+        $finder
+            ->files()
+            ->name($name)
+            ->ignoreUnreadableDirs()
+            ->in($directories)
+        ;
+
+        /** @var \SplFileInfo[] $finder */
+        foreach ($finder as $file) {
+            $filepaths[] = $file->getPathname();
+
+            if (true === $first) {
```

Comment:
This condition should only be checked once, before loop. If it is false, the check is unnecessary performed as many times as we have files.